### PR TITLE
Don't create pullbacks for function with not varied parameters

### LINKED ISSUE #682: Missing num-diff fallback warnings in nested function calls
Reproducer:
```c++
// Compile with clang++ -std=c++20 -fplugin=/usr/lib/clad.so missing_warning.cxx -o missing_warning

#include "clad/Differentiator/Differentiator.h"

#include <cmath>

inline double innerWrapper(double x)
{
   return std::lgamma(x); // call any function that needs num. diff.
}

double wrapper(double *params)
{
   // When calling the num-diffed function directly, you get the correct
   // warning:
   // return std::lgamma(params[0]);
   return innerWrapper(params[0]);
}

int main()
{
   auto grad = clad::gradient(wrapper, "params");
}
```

When calling `std::lgamma` directly, you get this correct warning when compiling:
```txt
warning: Falling back to numerical differentiation for 'lgamma' since no suitable overload was found and clad could not derive it. To disable this feature, compile your programs with -DCLAD_NO_NUM_DIFF.
1 warning generated.
```
However, the warning is absent in the reproducer with the nested function call.

It would be nice if you also get the warning in that situation, because these warnings are very useful to spot residual num. diffs in the gradients.

### PR DESCRIPTION
**Pull request by ovdiiuv**
This PR enables clad not to create pullbacks if the parameters are either constant or not varied. 

We should revert the changes once Varied Analysis is redesigned to track varied expression/statements and not only declarations. 

Fixes: #642, Fixes #682

### CODE REVIEW

File: lib/Differentiator/ActivityAnalyzer.cpp [hunk 1]
```diff
@@ -122,15 +122,15 @@ bool VariedAnalyzer::VisitCallExpr(CallExpr* CE) {
   bool noHiddenParam = (CE->getNumArgs() == FD->getNumParams());
   if (noHiddenParam) {
     MutableArrayRef<ParmVarDecl*> FDparam = FD->parameters();
-    m_Varied = true;
-    m_Marking = true;
+    // m_Varied = true;
+    // m_Marking = true;
```
Comment (vgvassilev) [inline lib/Differentiator/ActivityAnalyzer.cpp:126]: Can you remove the commented out code?

File: lib/Differentiator/ActivityAnalyzer.cpp [hunk 2]
```diff
@@ -141,7 +141,8 @@ bool VariedAnalyzer::VisitDeclStmt(DeclStmt* DS) {
       m_Varied = false;
       TraverseStmt(init);
       m_Marking = true;
-      if (m_Varied)
+      if (m_Varied || cast<VarDecl>(D)->getType()->isPointerType() ||
+          cast<VarDecl>(D)->getType()->isArrayType())
```
Comment (vgvassilev) [inline lib/Differentiator/ActivityAnalyzer.cpp:145]:
```suggestion
      QualType VDTy = cast<VarDecl>(D)->getType();
      if (m_Varied || isArrayOrPointerType(VDTy))
```

File: lib/Differentiator/DiffPlanner.cpp [hunk 3]
```diff
@@ -620,11 +620,20 @@ namespace clad {
...
+      std::vector<ParmVarDecl*> derivedParam;
+
+      for (auto* parameter : FDparam) {
+        QualType parType = parameter->getType();
+        if (parType->isPointerType()) {
+          if (!parType->getPointeeType().isConstQualified())
+            derivedParam.push_back(parameter);
+        } else if (!parType.isConstQualified())
+          derivedParam.push_back(parameter);
+      }
```
Comment (vgvassilev) [inline lib/Differentiator/DiffPlanner.cpp:634]:
```suggestion
      for (ParmVarDecl* Param : FDparam) {
        // Ignore const qualified types.
        QualType ParamTy = Param->getType();
        do {
           ParamTy = ParamTy->getPointeeType();
        } while(ParamTy->isPointerType());
        if (!ParamTy.isConstQualified())
          derivedParam.push_back(Param);
      }
```
Should cover cases such as `const*const**`.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 4]
```diff
@@ -1648,6 +1648,49 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+        bool isConst = clad_compat::Expr_EvaluateAsConstantExpr(subexprE, dummy,
+                                                                m_Context);
+        if (isConst)
+          continue;
+        if (const auto* ASE = dyn_cast<ArraySubscriptExpr>(subexprE)) {
+          const Expr* baseExpr = ASE->getBase()->IgnoreImpCasts();
+          auto* paramVD =
+              dyn_cast<VarDecl>(dyn_cast<DeclRefExpr>(baseExpr)->getDecl());
+          if (m_DiffReq.shouldHaveAdjoint(paramVD))
+            return true;
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1671]: Shouldn't we call `isVariedParam(subexpr)`?

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 5]
```diff
@@ -1648,6 +1648,32 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     return StmtDiff(clonedDRE);
   }
 
+  bool ReverseModeVisitor::isVariedParam(const Expr* paramE) {
+    if (isa<BinaryOperator>(paramE)) {
```
[the remark this answers is not in the document]
Comment (ovdiiuv) [reply lib/Differentiator/ReverseModeVisitor.cpp:1652]: yes

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 6]
```diff
@@ -1648,6 +1648,32 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+        return true;
+    } else if (const auto* DRE =
+                   dyn_cast<DeclRefExpr>(paramE->IgnoreImpCasts())) {
+      auto* paramVD = dyn_cast<VarDecl>(DRE->getDecl());
+      if (m_DiffReq.shouldHaveAdjoint(paramVD))
+        return true;
+    } else {
+      return true;
+    }
+    return false;
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1674]:
```suggestion
    class Collector : public RecursiveASTVisitor<Collector> {
      const DiffRequest & Request;
       Collector(const DiffRequest & DR) : Request(DR) {}
      bool IsVaried(Expr *E) {
        return !Visit(E); // false means it is varied
      }
      bool VisitDeclRefExpr(DeclRefExpr* DRE) {
        if (m_DiffReq.shouldHaveAdjoint(DRE->getDecl())
          return false; // Abort further visitation we got our answer.
        return true;
      }
    } DRECollector(m_DiffReq);
    return DRECollector->isVaried(paramE);
```

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 7]
```diff
@@ -1648,6 +1648,34 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     return StmtDiff(clonedDRE);
   }
 
+  class VariedChecker : public RecursiveASTVisitor<VariedChecker> {
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1651]: Can you move the class definition in the body of the function. It makes it clear that’s used only there.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 8]
```diff
@@ -1800,7 +1800,36 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
...
+              containsTargetVar = true;
+              return false;
+            }
+            return true;
+          }
+
+        private:
+          bool containsTargetVar = false;
+          const DiffRequest& Request;
+        };
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1830]:
```suggestion
        class VariedChecker : public RecursiveASTVisitor<VariedChecker> {
        const DiffRequest& Request;

        public: 
          VariedChecker(const DiffRequest& DR) : Request(DR) {}

          bool isVariedE(const clang::Expr* E) {
            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
            return TraverseStmt(const_cast<clang::Expr*>(E));
          }

          bool VisitDeclRefExpr(const clang::DeclRefExpr* DRE) {
            if (!isa<VarDecl>(DRE->getDecl()))
               return true;

            if (Request.shouldHaveAdjoint(cast<VarDecl>(DRE->getDecl())))
              return false; // varied, no need to descent more
            return true;
          }
          } analyzer(m_DiffReq); 
```

**Review by vgvassilev — APPROVED** [review]
This is great! Lgtm! Thank you!

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 9]
```diff
@@ -1800,7 +1800,24 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
         // subexpression.
         if (const auto* MTE = dyn_cast<MaterializeTemporaryExpr>(arg))
           arg = clad_compat::GetSubExpr(MTE)->IgnoreImpCasts();
-        if (!arg->isEvaluatable(m_Context)) {
+        class VariedChecker : public RecursiveASTVisitor<VariedChecker> {
```
Comment (vgvassilev) [inline lib/Differentiator/ReverseModeVisitor.cpp:1803]:
```suggestion
        // FIXME: We should consider moving this code in the VariedAnalysis
        // where we could decide to remove pullback requests from the
        // diff graph.
        class VariedChecker : public RecursiveASTVisitor<VariedChecker> {
```

**Pull request merged**